lib/rpc/SingleJsonRpcProvider.ts

```diff
@@ -128,29 +129,58 @@ export class SingleJsonRpcProvider extends StaticJsonRpcProvider {
     return super.getBlockNumber()
   }
 
-  private wrappedFunctionCall(fnName: string, fn: (...args: any[]) => Promise<any>, ...args: any[]): Promise<any> {
+  private async wrappedFunctionCall(
+    fnName: string,
+    fn: (...args: any[]) => Promise<any>,
+    ...args: any[]
+  ): Promise<any> {
     this.log.debug(`SingleJsonRpcProvider: wrappedFunctionCall: fnName: ${fnName}, fn: ${fn}, args: ${[...args]}`)
     const perf: SingleCallPerf = {
       succeed: true,
       latencyInMs: 0,
       startTimestampInMs: Date.now(),
     }
-    return fn(...args)
-      .then((response: any) => {
-        return response
-      })
-      .catch((error: any) => {
-        perf.succeed = false
-        this.log.error(JSON.stringify(error))
-        throw error
-      })
-      .finally(() => {
-        perf.latencyInMs = Date.now() - perf.startTimestampInMs
-        this.checkLastCallPerformance(fnName, perf)
-        this.lastCallTimestampInMs = perf.startTimestampInMs
+    try {
+      return await fn(...args)
+    } catch (error: any) {
+      perf.succeed = false
+      this.log.error(JSON.stringify(error))
+      throw error
+    } finally {
+      perf.latencyInMs = Date.now() - perf.startTimestampInMs
+      this.checkLastCallPerformance(fnName, perf)
+      this.updateHealthyStatus()
+      this.lastCallTimestampInMs = perf.startTimestampInMs
+
+      if (this.enableDbSync) {
+        this.maybeSyncHealthScoreAndUpdateHealthyStatus()
```

Comment:
if we do this only on the single provider on method invocation, doesn't this mean inactive providers will have stale health scores until selected and called?

Reply:
Currently we don't do shadow query on inactive provider, so we wouldn't know its health status. This is a trade off between awareness and number of wasted API calls.

But you are right, this is actually the target of my next PR. I will add the support for making shadow requests to poke the status for inactive providers.